### PR TITLE
common prov/efa: add ibv_fork_init check and use it in EFA provider

### PR DESCRIPTION
**Pull request by rwespetal**
Detect whether ibv_fork_init() has already been called during domain
initialization. If it is and the cache is enabled return an error.

Signed-off-by: Robert Wespetal <wesper@amazon.com>

Test description: Verified that an error is returned when the Open MPI verbs BTL calls ibv_fork_init() before Libfabric is loaded. Tested disabling the verbs BTL and the check passes as the ibv_fork_init call is avoided. Also tried the `RDMAV_FORK_SAFE` environment variable and that's also detected correctly.

### CODE REVIEW

**Comment by rwespetal** [comment]
Work in progress as I need to come up with a good error message for this case. The rest of it is ready to review though.

**Comment by shefty** [comment]
Can we move this functionality into a common location that efa and verbs can share?

I would also argue that this is highly dependent on the internal implementation of what verbs does, so we could also just go directly to the provider calls to bypass the internal fork handling.  But that would require more work.  (I had started down that path before getting distracted, and I don't think it's too bad.  We just need to be clever tracking fork support, so that we don't end up with memory alloc/free calls in the monitor handlers).

**Comment by rwespetal** [comment]
Sure, I'd be happy to do that.

I agree that this is highly dependent on rdma-core internal implementation. I was hoping they documented this explicit behavior in the manpage but unfortunately they don't. I suspect this code will not change though and I am hoping we can remove this sooner rather than later.

We do have applications that require fork() support -- ML/NCCL apps. NCCL uses bounce buffers and does its own registrations so it doesn't require a registration cache. It does call fork though so I think we can't bypass and go directly to the provider calls until we call madvise() in Libfabric for applications that need it.

File: prov/efa/src/efa_domain.c
Comment (wzamazon) [inline prov/efa/src/efa_domain.c:273]: Do we want to print some warning message here?

Comment (rwespetal) [reply prov/efa/src/efa_domain.c:273]: Yep, I haven't come up with an a good one yet. :)

Comment (wzamazon) [reply prov/efa/src/efa_domain.c:273]: sorry, missed your first comment. 

We also need to update the error message in ```efa_atfork_callback``` too. Currently we advise user to set RDMAV_FOR_SAFE, which won't work.

Comment (rwespetal) [reply prov/efa/src/efa_domain.c:273]: Ah, good point, thanks. I will update that comment too.

**Comment by rwespetal** [comment]
AWS CI failure is real. I'll have to update the scripts so we build Open MPI without verbs support.

**Comment by tschuett** [comment]
How about adding `ibv_fork_was_initialized()` to rdma-core?

**Comment by wzamazon** [comment]
I think we can also add ```--mca mtl ^openib``` to mpirun command line.

**Comment by bwbarrett** [comment]
> How about adding `bool ibv_fork_was_initialized()` to rdma-core to check whether ibv_fork_init was called before?

This would be the absolute wrong thing to do.  This whole patch is a hack.  A necessary hack, given the situation with rdma-core, but a hack.  There are use cases that need both fork and registration caches.  So rather than adding a "was initialized?" function, it would be better to fix the core issue.  Also, one advantage of this hack is that it breaks someone successfully enabling fork support later.

We should talk to the rdma-core team about building an approach that works for situations like these, but it should be more of a solving the problem approach and less of a make this patch cleaner approach.

File: prov/efa/src/efa_domain.c
Comment (zhngaj) [inline prov/efa/src/efa_domain.c:252]: I think I understand the first half statement, which is because the RBTree is not updated correctly in rdma-core if doing de-registration after the regions are unmapped in fork support enabled case. 

But not quite follow `deregistering regions after they are unmapped is required for Libfabric memory monitor intercept functions to work`. I thought we still want the immediately flush for memhooks monitor, there's just an issue in memhooks interception implementation recently merged. Could you clarify further? Sorry for missing anything.

Comment (rwespetal) [reply prov/efa/src/efa_domain.c:252]: We won't be able to fix the recently discovered issue with memhooks interception.

I'll remove this so there isn't confusion about it in the future.

**Comment by shefty** [comment]
My understanding is that the fork handling libibverbs causes one issue (fails memory deregistration); however, there's a second issue, which is calling free from within an intercept handler.  The latter problem is still there even if libibverbs fork handling is disabled.  The result is that we can't make any calls that lead to calling free from the intercepts, which includes calling ibv_dereg_mr.

File: include/ofi.h
Comment (wzamazon) [inline include/ofi.h:377]: I feel the verb ```check``` should be moved up, like```ofi_check_vrb_fork_enabled```. Also ```vrb``` is not clear to me, may be 'ibv'?

**Comment by wzamazon** [comment]
nccl test fails are real issues, and need to be looked into.

File: include/ofi.h
Comment (rwespetal) [reply include/ofi.h:377]: I was following the other verbs specific function in common -- `ofi_vrb_speed`.

**Comment by rwespetal** [comment]
> nccl test fails are real issues, and need to be looked into.

Yes, PR #6308 needs to go in first.

File: include/ofi.h
Comment (wzamazon) [reply include/ofi.h:377]: I see, thanks!

**Comment by rwespetal** [comment]
> My understanding is that the fork handling libibverbs causes one issue (fails memory deregistration); however, there's a second issue, which is calling free from within an intercept handler. The latter problem is still there even if libibverbs fork handling is disabled. The result is that we can't make any calls that lead to calling free from the intercepts, which includes calling ibv_dereg_mr.

Yes, this is all correct. We've backed out the changes to deregister in the intercept handlers because of the free() calls in ibv_dereg_mr. We know that fork support breaks when deregistering after the region is unmapped, we're not aware of any other issues in libibverbs or the EFA driver that prevents us from doing this as a short term fix.

**Comment by wzamazon** [comment]
> Yes, PR #6308 needs to go in first.

Thanks! I would prefer that we rerun the test after 6308 is merged then.

**Comment by rwespetal** [comment]
> > Yes, PR #6308 needs to go in first.
> 
> Thanks! I would prefer that we rerun the test after 6308 is merged then.

Of course, we'll wait to merge this until CI passes.

**Comment by wzamazon** [comment]
> however, there's a second issue, which is calling free from within an intercept handler.

I think what we confirmed is that calling ```malloc``` from intercept handler will caused dead lock on certain version of glibc. I do not think we have confirmed calling ```free``` would cause dead lock.

**Comment by bwbarrett** [comment]
> > however, there's a second issue, which is calling free from within an intercept handler.
> 
> I think what we confirmed is that calling `malloc` from intercept handler will caused dead lock on certain version of glibc. I do not think we have confirmed calling `free` would cause dead lock.

It's the same lock, and we know that free can try to acquire that lock.  So we should avoid calling either malloc or free in the callback from the memory hook.

**Review by shefty — APPROVED** [review]

**Comment by rwespetal** [comment]
bot:aws:retest

**Comment by shefty** [comment]
See: https://github.com/ofiwg/libfabric/issues/6327

This change breaks libfabric.  It was caught by our daily CI validation, not our per PR CI.

**Pull request merged**
